### PR TITLE
Update discoverers to conform to new naming schemes

### LINKED ISSUE #71: OpenStack discoverer: Improve handling of LB and Listener names
We use the LBaaS listener name as the port name. When the name is long enough, it can result in an invalid port name.

```
time="2018-04-26T14:54:23Z" level=error msg="error handling update endpoints 'vioadmin/1-61fb2abd-f007-4dd2-8784-9db807b27e4d-openstack': Endpoints "1-61fb2abd-f007-4dd2-8784-9db807b27e4d-openstack" is invalid: subsets[1].ports[0].name: Invalid value: "k8s-registry-listener-a2fd7571-5091-4a1f-8a65-610250ce8877-11000": must be no more than 63 characters"
```

### PR DESCRIPTION
**Pull request by alexbrand**
Fixes #71 

We arrived at a new naming scheme for discovered services in #102. This PR updates the discoverers to conform to these new naming conventions.

1. The discovered name is now `${backend-name}-${service-name}`.
1. The discovered name shortening algorithm now hashes each component individually, instead of taking the hash of the entire discovered name.
1. The OpenStack discoverer skips any load balancers which has a name that contains invalid characters or ends with a character other than a letter or number. Skipped LBs are logged and a metric is incremented.
1. The OpenStack discoverer no longer uses listener names in the discovered service port names.
1. Labels added during discovery are shortened if necessary.

### CODE REVIEW

**Comment by rosskukulinski** [comment]
and relatedly, should labels applied be updated? `gimbal.heptio.com/cluster` --> `gimbal.heptio.com/backend`?

**Comment by alexbrand** [comment]
Whoops! Looks like I have a typo in the description :(. Will fix.

**Comment by alexbrand** [comment]
Updating the labels makes sense to me. Will push commit shortly.

File: discovery/design/kubernetes.md [hunk 1]
```diff
@@ -76,7 +76,7 @@ If additional processing is required, then an argument can be changed to increas
 
 Those labels are defined as:
 
-- gimbal.heptio.com/Cluster: ClusterName (Defined via argument)
+- gimbal.heptio.com/backend: ClusterName (Defined via argument)
```
Comment (stevesloka) [inline discovery/design/kubernetes.md]: I think we should lowercase all the labels. Backend is now lower but the other's arent.

File: discovery/svc.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+apiVersion: v1
```
Comment (stevesloka) [inline discovery/svc.yaml]: This a rogue file?

File: discovery/design/kubernetes.md (hunk 1 above)
Comment (stevesloka) [reply discovery/design/kubernetes.md]: I think the code is fine just the docs have caps.

**Comment by alexbrand** [comment]
Thanks @stevesloka! Added missing docs/links.

**Comment by rosskukulinski** [comment]
From a product PoV, LGTM.  Waiting on @stevesloka for any final code review notes.

**Pull request merged**